### ISSUE TITLE
Make test failed. Is there anything wrong?

### ISSUE THREAD

**Issue by naveen126** [snps-virtualprototyping/ocx-qemu-arm#4]
Hi Sir, 
I have cloned a copy of this git of master branch.
I have folloed the steps to build it, but when I run "make test", I have got some failed information.

Running tests...
Test project /home/karl.zhao/Documents/ZooDevelopmentProject/UniCorn/ocx-qemu-arm/BUILD_OUT
    Start 1: ocx-qemu-arm
1/1 Test #1: ocx-qemu-arm .....................***Failed    0.05 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.05 sec

The following tests FAILED:
	  1 - ocx-qemu-arm (Failed)
Errors while running CTest
Output from these tests are in: /home/karl.zhao/Documents/ZooDevelopmentProject/UniCorn/ocx-qemu-arm/BUILD_OUT/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
make: *** [test] Error 8

I don't know what is wrong.
Does it need some dependency on centos 7?

**Comment by nbosb** [comment]
Hello,

the test fails because of the breakpoint behavior of the armcore. When the armcore hits a breakpoint, a (user-defined) callback function is called. This callback has the return type `bool`. When `true` is returned from the callback, a stop flag is set in the unicorn model ([see here](https://github.com/snps-virtualprototyping/ocx-qemu-arm/blob/master/src/armcore.cpp#L960)). In this case, the test expects that the unicorn TCG exits and the `step` function returns immediately. Instead, the instruction where the breakpoint is placed is still executed before the TCG exits. That results in a PC that is by 4 larger than expected which lets the test fail.

To fix this behavior, I created a patch for unicorn which can be found [here](https://github.com/aut0/avp64/blob/master/patches/unicorn-fix-breakpoint.patch). The patch stops the TCG before the instruction where the breakpoint is placed is executed. This patch only works for the case where the callback function returns `true` to stop the code execution. When the callback returns `false`, the simulation stucks at the breakpoint and calls the callback multiple times. For that case, the unpatched version works fine.

To summarize: The test fails because of the breakpoint behavior. If the breakpoint callback function always returns `false`, the current behavior is correct. If the callback function always returns `true`, my patch can be used. If the return value can be both, `true` and `false`, a new solution needs to be developed.
